src/main.rs

```diff
@@ -1,28 +1,177 @@
+//
+//  - If reverse proxy (runs for each new connection):
+//      - Check to see if payload includes TLS records, if not, forward the packet without modification
+//      - Create a new TLS server session (assuming rustls)
+//      - Create a new TCP connection to the target server
+//      - Loop until either client or server connection closes:
+//          - If data received from client, read the data using read_tls() and then read(), then send to server (assuming rustls)
+//          - If data received from server, send to client using write() and then write_tls() (assuming rustls)
+//              - Optionally compress data before sending
+
+mod forward_proxy;
+mod reverse_proxy;
+
+use error_chain::bail;
+mod errors {
+    error_chain::error_chain! {}
+}
+use errors::*;
+
+use clap::{Arg, App, SubCommand, AppSettings};
+use std::net::{TcpListener, IpAddr, SocketAddr};
+
+enum Server {
+    Forward { 
+        port: u16 
+    },
+    Reverse { 
+        port: u16, 
+        server_ips: Vec<SocketAddr> 
+    },
+}
+
+const APP_NAME : &str = "Rust TLS Proxy";
+const ABOUT_STR : &str = "Project for network systems class to build a \
+    transport TLS proxy in Rust to encrypt unencrypted messages";
+
+const FORWARD_PORT_HELP : &str = const_format::formatcp!(
+    "port number receiving intercepted client connections, default {}", 
+    forward_proxy::DEFAULT_PORT
+);
+
+const REVERSE_PORT_HELP : &str = const_format::formatcp!(
+    "port number receiving incoming connections, default {}",
+    reverse_proxy::DEFAULT_PORT
+);
+
+fn run() -> Result<()> {
+    let m = App::new(APP_NAME)
+        .about(ABOUT_STR)
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .arg(Arg::with_name("compress")
+             .short("c")
+             .long("compress")
+             .help("enable compression"))
+        .arg(Arg::with_name("encrypt")
+             .short("e")
+             .long("encrypt")
+             .help("enable encryption"))
+        .subcommands( vec![ 
+            SubCommand::with_name("forward")
+                .about("start in foward proxy server mode")
+                .arg(Arg::with_name("port")
+                     .short("p")
+                     .long("port")
+                     .help(FORWARD_PORT_HELP)
+                     .takes_value(true)),
+
+            SubCommand::with_name("reverse")
+                .about("start in reverse proxy server mode")
+                .arg(Arg::with_name("port")
+                     .short("p")
+                     .long("port")
+                     .help(REVERSE_PORT_HELP)
+                     .takes_value(true))
+                .arg(Arg::with_name("SERVERS")
+                     .help("server addresses in format ip:port")
+                     .required(true)
+                     .multiple(true))
+        ])
+        .get_matches_safe()
+        .chain_err(|| "error parsing arguments")?;
+
+    let compress = m.is_present("compress");
```

Comment:
Might be nice (but not critical) to add a defaults section with default values so that the compress assignment could be something like: 
```
let compress = match m.is_present("compress") {
    true => true,
    false => default["compress"];
};
```

That way we could set the defaults in a single place. I'm sure there's probably a more elegant way to do it than my above snippet.